aspnetcore/performance/blazor/webassembly-best-practices.md

```diff
@@ -0,0 +1,154 @@
+---
+title: ASP.NET Core Blazor WebAssembly performance best practices
+author: pranavkm
+description: Tips for increasing performance in ASP.NET Core Blazor WebAssembly apps and avoiding common performance problems.
+monikerRange: '>= aspnetcore-2.1'
+ms.author: riande
+ms.custom: mvc
+ms.date: 05/13/2020
+no-loc: [Blazor, "Identity", "Let's Encrypt", Razor, SignalR]
+uid: performance/blazor/webassembly-best-practices
+---
+# ASP.NET Core Blazor WebAssembly performance best practices
+
+By [Pranav Krishnamoorthy](https://github.com/pranavkm)
+
+This article provides guidelines for ASP.NET Core Blazor WebAssembly performance best practices.
+
+## Avoid unnecessary component renders
+
+Blazor's diffing algorithm avoids rerendering a component when the algorithm perceives that the component hasn't changed. Override [ComponentBase.ShouldRender](xref:Microsoft.AspNetCore.Components.ComponentBase.ShouldRender%2A) for fine-grained control over component rendering.
+
+If authoring a UI-only component that never changes after the initial render, configure `ShouldRender` to return `false`:
+
+```razor
+@code {
+    protected override bool ShouldRender() => false;
+}
+```
+
+Also use <xref:Microsoft.AspNetCore.Components.ComponentBase.ShouldRender%2A> to selectively render a component responding to a UI event.
+
+In the following example:
+
+* <xref:Microsoft.AspNetCore.Components.ComponentBase.ShouldRender%2A> is overridden and set to the value of the `shouldRender` field, which is initially `false` when the component loads.
+* When the button is selected, `shouldRender` is set to `true`, which forces the component to rerender with the updated `currentCount`.
+* Immediately after rerendering, <xref:Microsoft.AspNetCore.Components.ComponentBase.OnAfterRender%2A> sets the value of `shouldRender` back to `false` to prevent further rerendering until the next time the button is selected.
```

Comment:
@pranavkm I feel like this example is confusing. Because after reading this I've got a natural question whether I should do this myself going forward - as this seem to be very optimal.
We should add an explicit statement saying that this particular example wouldn't need such feel in practice, as the diffing algorithm would get to the same result and that his has been done only for demonstration purposes.

Reply:
I was shooting for do this if you want fine grained control over when you want to be rendered. Is there a better way to phrase this @mkArtakMSFT  \ @guardrex ?

Reply:
Maybe dissuade a bit in the lead-in.

Currently, it starts with ...

> Also use `ShouldRender` to selectively render a component responding to a UI event.

... but it could dissuade a bit and say something like ...

> Most apps won't require fine-grained control, but `ShouldRender` can also be used to selectively render a component responding to a UI event.

Reply:
Yeah ... let's go with that. I'll shoot it in now.